mypy/subtypes.py

```diff
@@ -1007,32 +1007,41 @@ def unify_generic_callable(type: CallableType, target: CallableType,
 
 
 def restrict_subtype_away(t: Type, s: Type, *, ignore_promotions: bool = False) -> Type:
-    """Return t minus s.
+    """Return t minus s for runtime type assertions.
 
     If we can't determine a precise result, return a supertype of the
     ideal result (just t is a valid result).
 
     This is used for type inference of runtime type checks such as
-    isinstance.
-
-    Currently this just removes elements of a union type.
+    isinstance(). Currently this just removes elements of a union type.
     """
     if isinstance(t, UnionType):
-        # Since runtime type checks will ignore type arguments, erase the types.
-        erased_s = erase_type(s)
-        # TODO: Implement more robust support for runtime isinstance() checks,
-        # see issue #3827
         new_items = [item for item in t.relevant_items()
-                     if (not (is_proper_subtype(erase_type(item), erased_s,
-                                                ignore_promotions=ignore_promotions) or
-                              is_proper_subtype(item, erased_s,
-                                                ignore_promotions=ignore_promotions))
-                         or isinstance(item, AnyType))]
+                     if (isinstance(item, AnyType) or
+                         not covers_at_runtime(item, s, ignore_promotions))]
         return UnionType.make_union(new_items)
     else:
         return t
 
 
+def covers_at_runtime(item: Type, supertype: Type, ignore_promotions: bool) -> bool:
+    """Will isinstance(item, supertype) always return True at runtime?"""
+    # Since runtime type checks will ignore type arguments, erase the types.
+    supertype = erase_type(supertype)
+    if is_proper_subtype(erase_type(item), supertype, ignore_promotions=ignore_promotions):
+        return True
+    if isinstance(supertype, Instance) and supertype.type.is_protocol:
```

Comment:
Why do we need the `is_protocol` check? It wasn't there previously.

Reply:
This whole branch was initially added as a crutch to support `isinstance()` with protocols (see the TODO I moved). The point is that `erase_type()` only erases type arguments but not attribute types. (Well, the check is still unsafe, this is why we have the issue).

Also for non-protocol classes this branch should be a no-op (unless there is a bug in `is_proper_subtype()`).